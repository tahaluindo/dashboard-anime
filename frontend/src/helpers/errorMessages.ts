enum errors {
  NOT_AUTHORIZED = "Usuário não autorizado a realizar essa operação.",
  SERVER_COMMUNICATION = "Erro ao comunicar com servidor.",
  OPERATION_ERROR = "Erro ao realizar operação.",
}

export default errors;

interface CODE {
  code?: string,
}

interface ERROR {
  extensions?: CODE,
}

export function forbidenError(errors: ERROR[] | undefined): boolean {
  if (errors?.shift()?.extensions?.code === "forbidden") return true;

  return false;
}