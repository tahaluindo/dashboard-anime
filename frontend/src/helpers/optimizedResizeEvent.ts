class OptimizedResizeEvent {
  private callbacks: Function[] = [];
  private running: boolean = false;

  constructor() {
    this.resize.bind(this);
    this.runCallbacks.bind(this);
  }

  private runCallbacks() {
    this.callbacks.forEach((callback) => callback());
    this.running = false;
  }

  private resize() {
    if (!this.running) {
      this.running = true;
      setTimeout(
        () => window.requestAnimationFrame(this.runCallbacks.bind(this)),
        500
      );
    }
  }

  private addCallback(callback: Function) {
    this.callbacks.push(callback);
  }

  add(callback: Function) {
    if (!this.callbacks.length)
      window.addEventListener("resize", this.resize.bind(this));
    this.addCallback(callback);
  }
}

export const optimizedResizeEvent = new OptimizedResizeEvent();
