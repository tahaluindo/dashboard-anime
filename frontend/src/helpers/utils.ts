import { inject, InjectionKey } from "vue";

export function numberToReal(number: number) {
  return Intl.NumberFormat("pt-BR", {
    style: "currency",
    currency: "BRL",
  }).format(number);
}

export function isMobile() {
  return window.visualViewport.width < 1000;
}

export function getGridColumnsQty() {
  const windowSize = window.visualViewport.width;

  if (windowSize < 768) return 1;
  if (windowSize < 1024) return 2;
  if (windowSize < 1280) return 3;
  return 4;
}

export function onClickOutside(
  target: HTMLElement,
  self: HTMLElement
): boolean {
  const compareWithThisAndParent = (element: Element) =>
    element === target || element === target.parentElement;
  const isChild =
    Array.from(self.children).findIndex(compareWithThisAndParent) !== -1;

  const isSelf = target === self;

  return !isSelf && !isChild;
}

export function isOdd(number: number) {
  return number % 2 !== 0;
}

export function madeOddArray<T>(arr: Array<T>): Array<T> {
  if (isOdd(arr.length)) return arr;

  return arr.slice(0, arr.length - 1);
}

export function getElement(element: HTMLElement | undefined): HTMLElement {
  if (!element) throw new Error("Element not setted!");

  return element;
}

export function injectStrict<T>(key: InjectionKey<T>, fallback?: T) {
  const resolved = inject(key, fallback);
  if (!resolved) throw new Error(`Could not resolve ${key.description}`);

  return resolved;
}
