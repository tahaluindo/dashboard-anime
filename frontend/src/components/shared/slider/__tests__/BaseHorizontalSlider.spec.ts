import { mount } from "@vue/test-utils";
import { nextTick } from "vue";
import { optimizedResizeEvent } from "../../../../helpers/optimizedResizeEvent";
import { mockTouchEvent } from "../../../../__mocks__/event.mock";
import BaseHorizontalSlider from "../BaseHorizontalSlider.vue";

describe("BaseHorizontalSlider component test", () => {
  it("should swap position when touch the screen.", async () => {
    const wrapper = mount(BaseHorizontalSlider, {
      props: {
        resizeEvent: optimizedResizeEvent,
        visibleElements: 1,
      },
      slots: {
        default:
          "<div style='width: 100px' class='test'>Test</div><div style='width: 100px'>Test</div><div style='width: 100px'>Test</div><div style='width: 100px'>Test</div><div style='width: 100px'>Test</div>",
      },
    });

    expect(wrapper.html()).toMatchSnapshot();

    const divElement = wrapper.find("div").element;
    const element = divElement.querySelector(".test") as HTMLElement;
    expect(element?.parentElement?.style.opacity).toBe("1");

    wrapper.vm.onTouchStart(mockTouchEvent(0));
    await nextTick();
    wrapper.vm.onTouchEnd(mockTouchEvent(100));
    await nextTick();

    expect(element?.parentElement?.style.opacity).toBe("1");

    wrapper.vm.onTouchStart(mockTouchEvent(100));
    await nextTick();
    wrapper.vm.onTouchEnd(mockTouchEvent(0));
    await nextTick();

    expect(divElement.querySelector("span")?.style.opacity).toBe("1");
  });

  it("should swap position when click the controller.", async () => {
    const wrapper = mount(BaseHorizontalSlider, {
      slots: {
        default: "<div>Test</div>",
      },
    });

    const divElement = wrapper.find("div").element;
    expect(divElement.querySelector("span")?.style.opacity).toBe("1");

    wrapper.vm.turn("right");
    await nextTick();

    expect(divElement.querySelector("span")?.style.opacity).toBe("0");
  });

  it("should disable controllers when visibleElements equals items length.", async () => {
    const wrapper = mount(BaseHorizontalSlider, {
      slots: {
        default: "<div>TEST</div>",
      },
      props: {
        visibleElements: 1,
      },
    });

    expect(wrapper.vm.enableControllers).toBeFalsy();

    const divElement = wrapper.find("div").element;
    expect(divElement.querySelector("span")?.style.opacity).toBe("1");

    wrapper.vm.onTouchStart(mockTouchEvent(0));
    await nextTick();
    wrapper.vm.onTouchEnd(mockTouchEvent(100));
    await nextTick();

    expect(divElement.querySelector("span")?.style.opacity).toBe("1");
  });
});
