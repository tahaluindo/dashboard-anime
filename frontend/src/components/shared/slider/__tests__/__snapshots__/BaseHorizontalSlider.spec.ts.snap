// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseHorizontalSlider component test should swap position when touch the screen. 1`] = `
"<section class=\\"
      grid
      items-center
      justify-center
      h-full
      grid-cols-1
      lg:grid-cols-slider
    \\"><span class=\\"
        left-0
        hidden
        cursor-pointer
        duration-450
        h-1/2
        top-1/4
        lg:block
        hover:duration-150
        text-true-gray-500
        hover:text-black
      \\"><svg width=\\"58.667\\" height=\\"101.033\\" viewBox=\\"0 0 58.667 101.033\\"><g transform=\\"matrix(0.999, 0.035, -0.035, 0.999, -520.017, -852.706)\\"><line x1=\\"34.075\\" y2=\\"39\\" transform=\\"translate(563.5 844.5)\\" fill=\\"none\\" stroke-linecap=\\"round\\" stroke-width=\\"15\\"></line><line x1=\\"34.075\\" y1=\\"39\\" transform=\\"translate(563.5 883.5)\\" fill=\\"none\\" stroke-linecap=\\"round\\" stroke-width=\\"15\\"></line></g></svg></span>
  <div class=\\"relative
        flex
        justify-start
        w-4/5
        h-full
        mx-auto
        overflow-hidden
        lg:w-full\\"><span style=\\"height: 100%; position: absolute; display: flex; justify-content: center; align-items: center; width: 0px; opacity: 1; transform: translateX(0px) ;\\"><div style=\\"width: 100px;\\" class=\\"test\\">Test</div></span><span style=\\"height: 100%; position: absolute; display: flex; justify-content: center; align-items: center; width: 0px; opacity: 0; transform: translateX(0px) ;\\"><div style=\\"width: 100px;\\">Test</div></span><span style=\\"height: 100%; position: absolute; display: flex; justify-content: center; align-items: center; width: 0px; transform: translateX(0px) ; opacity: 0;\\"><div style=\\"width: 100px;\\">Test</div></span><span style=\\"height: 100%; position: absolute; display: flex; justify-content: center; align-items: center; width: 0px; transform: translateX(0px) ; opacity: 0;\\"><div style=\\"width: 100px;\\">Test</div></span><span style=\\"height: 100%; position: absolute; display: flex; justify-content: center; align-items: center; width: 0px; transform: translateX(0px) ; opacity: 0;\\"><div style=\\"width: 100px;\\">Test</div></span></div><span class=\\"
        right-0
        hidden
        transform
        rotate-180
        cursor-pointer
        duration-450
        h-1/2
        top-1/4
        lg:block
        hover:duration-150
        text-true-gray-500
        hover:text-black
      \\"><svg width=\\"58.667\\" height=\\"101.033\\" viewBox=\\"0 0 58.667 101.033\\"><g transform=\\"matrix(0.999, 0.035, -0.035, 0.999, -520.017, -852.706)\\"><line x1=\\"34.075\\" y2=\\"39\\" transform=\\"translate(563.5 844.5)\\" fill=\\"none\\" stroke-linecap=\\"round\\" stroke-width=\\"15\\"></line><line x1=\\"34.075\\" y1=\\"39\\" transform=\\"translate(563.5 883.5)\\" fill=\\"none\\" stroke-linecap=\\"round\\" stroke-width=\\"15\\"></line></g></svg></span>
</section>"
`;
